### ISSUE TITLE
UnscopedRef can be different in interface implementation, leading to ref safety holes

### ISSUE THREAD

**Issue by jjonescz** [dotnet/roslyn#76100]
**Version Used**: 20241126.6 ([b23da04](https://github.com/dotnet/roslyn/commit/b23da0417d19a84b2032f886481ba0bb3c5def38))

**Steps to Reproduce**:

```cs
using System;
using System.Diagnostics.CodeAnalysis;

var r = F();
Console.WriteLine(r.F.ToString());

static R F()
{
    var r = new R();
    ((I)new C()).M(ref r);
    return r;
}

interface I
{
    void M(ref R r);
}

class C : I
{
    public void M([UnscopedRef] ref R r)
    {
        r.F = ref r.I;
    }
}

ref struct R
{
    public int I;
    public ref int F;
}
```

**Expected Behavior**: An error at `((I)new C()).M(ref r);` - an error would be reported there if replaced with `new C().M(ref r);`.

**Actual Behavior**: No errors. The program accesses an invalid reference at runtime.

**Comment by Sergio0694** [comment]
Just so I understand, the problem is just because the implementation is narrowing the escape scope, right? Ie. if you had:

```csharp
interface I
{
    void M([UnscopedRef] ref R r);
}

class C : I
{
    public void M(ref R r)
    {
    }
}
```

That should be fine, right? Would that still be allowed?

**Comment by jjonescz** [comment]
> That should be fine, right? Would that still be allowed?

Yes, that should be fine. Per [the spec](https://github.com/dotnet/csharplang/blob/56dae0a6e0efd3479fe55cd8eb6d6c2c0a53f4aa/proposals/csharp-11.0/low-level-struct-improvements.md#parameter-scope-variance), removing `[UnscopedRef]` in the interface implementation is allowed.

**Comment by jjonescz** [comment]
Also this section of the spec needs to be updated - it does not consider the "silly cyclic self-assignment":

> The compiler will report a diagnostic for _unsafe scoped mismatches_ across overrides, interface implementations, and delegate conversions when:
> - The method returns a `ref struct` or returns a `ref` or `ref readonly`, or the method has a `ref` or `out` parameter of `ref struct` type, and
> - The method has at least one additional `ref`, `in`, or `out` parameter, or a parameter of `ref struct` type.

**Comment by jjonescz** [comment]
Reopening as the PR was reverted.
